File: ieee/ieee754.go
package main

import (
    "encoding/binary"
    "fmt"
    "math"
    "os"
    "strconv"
    "strings"
)

// Convert float32 to IEEE-754 binary string
func ieee754ToBin(num float32) string {
    // Get the raw bits of the float32
    bits := math.Float32bits(num)
    binStr := make([]byte, 32)

    for i := 31; i >= 0; i-- {
        if (bits>>i)&1 == 1 {
            binStr[31-i] = '1'
        } else {
            binStr[31-i] = '0'
        }
    }
    return string(binStr)
}

// Convert IEEE-754 binary string to float32
func binToIEEE754(binStr string) (float32, error) {
    if len(binStr) != 32 {
        return 0, fmt.Errorf("binary string must be 32 bits")
    }

    var bits uint32
    for i := 0; i < 32; i++ {
        switch binStr[i] {
        case '1':
            bits |= 1 << (31 - i)
        case '0':
            // do nothing
        default:
            return 0, fmt.Errorf("invalid character in binary string")
        }
    }

    return math.Float32frombits(bits), nil
}

func main() {
    if len(os.Args) != 2 {
        fmt.Printf("Usage: %s <float_number>\n", os.Args[0])
        os.Exit(1)
    }

    // Parse input as float32
    input64, err := strconv.ParseFloat(os.Args[1], 32)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Invalid float: %v\n", err)
        os.Exit(1)
    }
    input := float32(input64)

    // Convert float to binary string
    binaryStr := ieee754ToBin(input)
    fmt.Printf("IEEE-754 binary of %.6f: %s\n", input, binaryStr)

    // Convert binary string back to float
    recovered, err := binToIEEE754(binaryStr)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Error: %v\n", err)
        os.Exit(1)
    }
    fmt.Printf("Recovered float from binary: %.6f\n", recovered)
}

File: ieee/ieee754.jl
# Convert Float32 to IEEE-754 binary string
function ieee754_to_bin(num::Float32)::String
    bits = reinterpret(UInt32, num)  # view the float's bits as UInt32
    bin_chars = Vector{Char}(undef, 32)
    for i in 31:-1:0
        bin_chars[32 - i] = ((bits >> i) & 0x1) == 1 ? '1' : '0'
    end
    return String(bin_chars)
end

# Convert IEEE-754 binary string to Float32
function bin_to_ieee754(binstr::AbstractString)::Float32
    if length(binstr) != 32
        error("Binary string must be 32 bits.")
    end
    bits::UInt32 = 0
    for (i, ch) in enumerate(binstr)
        if ch == '1'
            bits |= UInt32(1) << (32 - i)
        elseif ch != '0'
            error("Invalid character in binary string.")
        end
    end
    return reinterpret(Float32, bits)
end

# Main-like function
function main()
    if length(ARGS) != 1
        println("Usage: julia script.jl <float_number>")
        exit(1)
    end

    # Parse input as Float32
    input = parse(Float32, ARGS[1])

    # Convert float to binary string
    binary = ieee754_to_bin(input)
    @printf("IEEE-754 binary of %.6f: %s\n", input, binary)

    # Convert binary string back to float
    recovered = bin_to_ieee754(binary)
    @printf("Recovered float from binary: %.6f\n", recovered)
end

# Run if script is executed directly
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
